File: src/claudectl/domain/git.py
"""Low-level git operations for workspace management."""

from __future__ import annotations

from pathlib import Path

from git import InvalidGitRepositoryError, Repo
from git.exc import GitCommandError

from claudectl.domain.exceptions import NotInGitRepoError


def get_repo_root() -> Path:
    """Get the root directory of the current git repository.

    Correctly handles worktrees by finding the actual repository root
    instead of the worktree directory.

    Returns:
        Path to repository root

    Raises:
        NotInGitRepoError: If not in a git repository
    """
    try:
        repo = Repo(search_parent_directories=True)

        # For worktrees, repo.common_dir returns the shared .git directory
        # (which may contain relative path components like ..).
        # Resolve first to normalize, then get parent to get repo root.
        common_dir = Path(repo.common_dir).resolve()
        return common_dir.parent
    except InvalidGitRepositoryError as exc:
        raise NotInGitRepoError("Not in a git repository") from exc


def get_repo_name() -> str:
    """Get the name of the current git repository.

    Returns:
        Repository directory name
    """
    return get_repo_root().name


def get_current_branch() -> str | None:
    """Get the name of the current branch.

    Returns:
        Branch name, or None if in detached HEAD state
    """
    try:
        repo = Repo(search_parent_directories=True)
        if repo.head.is_detached:
            return None
        return repo.active_branch.name
    except (InvalidGitRepositoryError, TypeError):
        # TypeError can occur if head is detached
        return None


def branch_exists(branch_name: str) -> bool:
    """Check if a branch exists locally or remotely.

    Args:
        branch_name: Name of the branch to check

    Returns:
        True if branch exists, False otherwise

    Raises:
        NotInGitRepoError: If not in a git repository
    """
    try:
        repo = Repo(search_parent_directories=True)

        # Check local branches
        if branch_name in [head.name for head in repo.heads]:
            return True

        # Check remote branches (if origin exists)
        try:
            origin = repo.remotes.origin
            if branch_name in [ref.remote_head for ref in origin.refs]:
                return True
        except (AttributeError, IndexError):
            # No origin remote configured
            pass

        return False
    except InvalidGitRepositoryError as exc:
        raise NotInGitRepoError("Not in a git repository") from exc


def is_worktree_clean(worktree_path: Path) -> tuple[bool, str]:
    """Check if a worktree has uncommitted changes.

    Args:
        worktree_path: Path to the worktree to check

    Returns:
        Tuple of (is_clean, status_message)
    """
    try:
        repo = Repo(worktree_path)

        # Check for staged and unstaged changes
        if repo.is_dirty(untracked_files=False):
            # Count modified files
            modified_files = len(repo.index.diff(None)) + len(repo.index.diff("HEAD"))
            untracked_files = len(repo.untracked_files)

            parts = []
            if modified_files:
                parts.append(f"{modified_files} modified")
            if untracked_files:
                parts.append(f"{untracked_files} untracked")

            return False, ", ".join(parts)

        # Check for untracked files only
        if repo.untracked_files:
            return False, f"{len(repo.untracked_files)} untracked"

        return True, "Clean"

    except (InvalidGitRepositoryError, GitCommandError) as e:
        return False, f"Failed to check status: {e}"

File: internal/git/git.go
// Package git provides Git repository utilities and worktree management.
package git

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/plumbing"
)


// GetRepoRoot returns the root directory of the current git repository.
// Correctly handles worktrees by finding the actual repository root
// instead of the worktree directory.
func GetRepoRoot() (string, error) {
	wd, err := filepath.Abs(".")
	if err != nil {
		return "", err
	}
	return discoverRepoRoot(wd)
}

// GetRepoName returns the name of the current git repository.
func GetRepoName() (string, error) {
	root, err := GetRepoRoot()
	if err != nil {
		return "", err
	}
	return filepath.Base(root), nil
}

// GetCurrentBranch returns the name of the current branch.
// Returns empty string if in detached HEAD state.
// Correctly handles worktrees by detecting the branch from the current directory.
func GetCurrentBranch(path string) (string, error) {
	// Discover the actual repo root (handles worktrees)
	repoRoot, err := discoverRepoRoot(path)
	if err != nil {
		return "", err
	}

	// Check if we're in a worktree by looking for .git file
	gitDir := filepath.Join(path, ".git")
	if info, err := os.Stat(gitDir); err == nil && !info.IsDir() {
		// We're in a worktree - read HEAD from the worktree's git directory
		data, err := os.ReadFile(gitDir) //nolint:gosec // Reading .git file is safe, path is controlled
		if err == nil && len(data) > 8 && string(data[:8]) == "gitdir: " {
			worktreeGitDir := strings.TrimSpace(string(data[8:]))
			if !filepath.IsAbs(worktreeGitDir) {
				worktreeGitDir = filepath.Join(path, worktreeGitDir)
			}
			headFile := filepath.Join(worktreeGitDir, "HEAD")
			if headData, err := os.ReadFile(headFile); err == nil { //nolint:gosec // Reading HEAD file is safe, path is controlled
				headRef := strings.TrimSpace(string(headData))
				// Format: ref: refs/heads/branch-name or just commit hash
				if strings.HasPrefix(headRef, "ref: ") {
					headRef = strings.TrimPrefix(headRef, "ref: ")
					headRef = strings.TrimSpace(headRef)
				}
				if strings.HasPrefix(headRef, "refs/heads/") {
					return strings.TrimPrefix(headRef, "refs/heads/"), nil
				}
				// Detached HEAD
				return "", nil
			}
		}
	}

	// Regular repo or fallback - use go-git
	repo, err := OpenRepo(repoRoot)
	if err != nil {
		return "", err
	}

	head, err := repo.Head()
	if err != nil {
		return "", nil // Detached HEAD or no commits
	}

	if !head.Name().IsBranch() {
		return "", nil // Not on a branch
	}

	return head.Name().Short(), nil
}

// BranchExists checks if a branch exists locally or remotely.
func BranchExists(repoRoot, branchName string) (bool, error) {
	repo, err := OpenRepo(repoRoot)
	if err != nil {
		return false, err
	}

	// Check local branches
	branches, err := repo.Branches()
	if err != nil {
		return false, err
	}

	found := false
	err = branches.ForEach(func(ref *plumbing.Reference) error {
		if ref.Name().Short() == branchName {
			found = true
			return fmt.Errorf("found") // Break iteration
		}
		return nil
	})
	if err != nil && err.Error() == "found" {
		return true, nil
	}
	if found {
		return true, nil
	}

	// Check remote branches
	remotes, err := repo.Remotes()
	if err != nil {
		return false, err
	}

	for _, remote := range remotes {
		refs, err := remote.List(&git.ListOptions{})
		if err != nil {
			continue
		}
		for _, ref := range refs {
			if ref.Name().Short() == fmt.Sprintf("%s/%s", remote.Config().Name, branchName) {
				return true, nil
			}
		}
	}

	return false, nil
}
